timer/scheduler.go

```diff
@@ -0,0 +1,63 @@
+package timer
+
+import (
+	"time"
+
+	"github.com/nivista/steady/.gen/protos/common"
+	rcron "github.com/robfig/cron/v3"
+)
+
+type cron struct {
+	sched               rcron.Schedule
+	startTime, stopTime time.Time
+	maxExecutions       int
+}
+
+var parser = rcron.NewParser(rcron.Minute | rcron.Hour | rcron.Dom | rcron.Month | rcron.Dow | rcron.SecondOptional | rcron.Descriptor)
+
+// infiniteExecutions indicates a timer should fire until cancelled.
+const infiniteExecutions = 0
+
+func (c cron) schedule(prog Progress, now time.Time) (nextFire time.Time, done bool) {
+	if c.maxExecutions != infiniteExecutions && prog.CompletedExecutions >= c.maxExecutions {
+		done = true
+		return
+	}
+
+	if prog.LastExecution.IsZero() {
+		// subtract one second so sched.Next will return startTime if it is valid
+		nextFire = c.startTime.Add(-1 * time.Second)
```

Comment:
I don't follow why we are subtracting a second here?

The second should be a constant by the way with a comment.

Reply:
We subtract a second so that if startTime is a valid fire time, we will fire on startTime. Calling c.sched.Next on a valid fire time will go to the next valid fire time, we want to stay on the current valid fire time, it's a special case.

Reply:
ok